Update is_hermitian evaluation in Mul class to check the definition of an hermitian operator

mathiscr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #27988 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

In the Mul class, _eval_is_hermitian now checks the definition of an hermitian operator. $`A^\dagger = A`$. 

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * Added the check of the definition of an hermitian matrix
<!-- END RELEASE NOTES -->

mathiscr committed:
Enhance is_hermitian evaluation in Mul class

Is_hermitian now evaluate the definition, it tests if M = M.adjoint in
the Mul class.

mathiscr committed:
Change list comprehension to generator in all()

akhmerov commented:
The logic looks correct, but the fix is unlikely sufficient. I think sympy is limited when comparing noncommutative expressions to zero with `== 0`.

Check out this example:
```py
from sympy.physics.quantum.dagger import Dagger
from sympy.physics.quantum.boson import BosonOp
a = BosonOp('a')
expr = (3 * a + 4).adjoint() + (3 * a + 4)
assert (expr.adjoint() - expr) != 0
assert expr.adjoint() == expr
```

oscarbenjamin commented:
See the guide for three valued logic here:
https://docs.sympy.org/latest/guides/booleans.html

akhmerov commented:
Unfortunately, three-valued logic isn't the culprit here. The assertion `assert (expr.adjoint() - expr).is_zero() is False` also passes.

oscarbenjamin commented:
Ah, yes: `is_zero` will always be False for non-commutative expressions:

https://docs.sympy.org/latest/guides/assumptions.html#interpretation-of-the-predicates

Having commutativity as part of the assumptions system is awkward.

akhmerov commented:
Do you have a suggestion of how to improve the check for equality? I have encountered this problem in my own code recently.

oscarbenjamin commented:
Firstly the assumptions code here should not compute `a + a.adjoint()`. Creating new expressions within the old assumptions risks infinite recursion. Also these queries are potentially called as part of evaluating any symbolic expression so we should minimise what happens in them. Calling `.adjoint` like this is unacceptably expensive during normal symbolic evaluation:
```python
In [5]: x, y = symbols('x, y')

In [6]: x*y > 0
Mul._eval_is_hermitian called
Out[6]: x⋅y > 0
```
Generally I think that `is_hermitian` should not be part of the assumptions system at all. If you look at the list of assumptions predicates then it clearly does not belong there:
https://docs.sympy.org/latest/guides/assumptions.html#predicates

There could be a function `is_hermitian(expr)` or something but again it should probably not work by creating new expressions but rather examining the structure of the expression. Having an `is_` query depend on symbolic evaluation is usually backwards because you want to use `is_` queries during symbolic evaluation.

As for testing when a non-commutative expression is zero I don't know. I don't work with non-commutative expressions much. For matrices there is `is_zero_matrix`. Perhaps there should be some generalisation of `is_zero` that applies to all things that are not necessarily (extended) complex numbers.

TiloRC commented on sympy/core/mul.py:
```diff
@@ -1541,9 +1541,13 @@ def _eval_is_imaginary(self):
             return self._eval_real_imag(False)
 
     def _eval_is_hermitian(self):
+        if all((t1 - t2.adjoint()) == 0 for t1, t2 in zip(self.args, self.args[::-1])):
+            return True
```
Currently, for an expr like `A*B*C` the code checks if the following pairs are adjoint: (A, C), (B,B), (C, A). But checking (C, A) is redundant as its the same thing as checking (A, C).

So if you wanted to avoid repeated work you could do something like this:
```python
args = self.args
half = (len(args) + 1) // 2

if all((args[i] - args[-i-1].adjoint()) == 0
       for i in range(half)):
    return True
```

oscarbenjamin replied:
It should not use `.adjoint`.

New expressions should not be created as part of assumptions queries in the old assumptions.

oscarbenjamin commented:
Let's just make sure we are all clear about why the hermitian handler should not add arbitrary code. I just made a change to print out when this handler is called:
```diff
diff --git a/sympy/core/mul.py b/sympy/core/mul.py
index fd83c8610a..1096165d0f 100644
--- a/sympy/core/mul.py
+++ b/sympy/core/mul.py
@@ -1541,6 +1541,7 @@ def _eval_is_imaginary(self):
             return self._eval_real_imag(False)
 
     def _eval_is_hermitian(self):
+        print('Mul._eval_is_hermitian called for', self)
         return self._eval_herm_antiherm(True)
 
     def _eval_is_antihermitian(self):
```
Now let's run some completely unrelated operation:
```python
In [2]: diff(atanh(x*y), x)
Mul._eval_is_hermitian called for x*y
Mul._eval_is_hermitian called for -x**2*y**2
Mul._eval_is_hermitian called for y/(-x**2*y**2 + 1)
Out[2]: 
     y     
───────────
   2  2    
- x ⋅y  + 1

In [3]: diff(atanh(x), x)
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Mul._eval_is_hermitian called for -x**2
Out[3]: 
  1   
──────
     2
1 - x 
```
These handlers are called every time anything happens especially as part of creating new expressions. It is important that these handlers must not create new expressions because that runs a significant risk of infinite recursion and also generally makes everything slow.

Instead what should be done is to remove `is_hermitian` from the assumptions system completely. It should never have been added there. We can keep `is_hermitian` as a property but it should not be part of the assumptions system. Then if you want to make `is_hermitian` as slow as you like I don't mind.

TiloRC commented:
> Instead what should be done is to remove is_hermitian from the assumptions system completely

It would probably make sense to keep `Q.hermitian` and add this rule to the new assumptions. That way this sort of simplification could be done by calling refine.

mathiscr commented:
I have a question, if we remove `is_hermitian` from the assumptions system and keep `is_hermitian` as a property, then how to handle the cases : ?
```python
A = Operator("A")
A.is_hermitian = True
#or 
B = Operator("B", is_hermitian=True)
```

TiloRC commented:
The old assumption system is only able to handle predicates from [this](https://docs.sympy.org/latest/guides/assumptions.html#predicates) table. It's really weird that hermitian and antihermitian are predicates in that table as all of the other predicates are for scalars. I agree with Oscar that hermitian and antihermitian should probably be removed from the old assumptions. 

The new assumption system can currently handle some matrix inferences. However, its ability to reason about matrices should be improved, and maybe support for inferences about operators should be added.

---

The examples you gave don't actually currently work. 

```python
B = Operator("B", hermitian=True) # raises TypeError: QExpr._eval_args() got an unexpected keyword argument 'hermitian'
```

And the following works because it's possible to define arbitrary new attributes to operators.
```python
A = Operator("A")
A.is_hermitian = True
print(A.is_hermitian) # True

A.blah = True
print(A.blah) # True
```

I don't think that's intentional. Operators should probably raise attribute errors in this sort of situation like Symbols do:
```python
from sympy import symbols

x = symbols("x")
x.is_real = True # raises AttributeError
print(x.blah) # raises AttributeError
x.blah = True # raises AttributeError
```

mathiscr committed:
Remove assumptions for hermitian and antihermitian

- Remove hermitian and antihermitian from the assomptions system
- Keep is_hermitian and is_antihermitian as properties
- Implement the new is_hermitian and is_antihermitian in the classes Expr, Add, Mul, Pow, Operator, HermitianOperator

oscarbenjamin commented:
> I removed hermitian and antihermitian from the assumptions system

If you want to do that then I think it is important for it to be a separate PR from making any other changes to how is_hermitian is calculated.

> For example, x = symbols("x", zero=True) implies that x is Hermitian.

You can add an `is_hermitian` to Symbol or Expr that checks this like:
```
if self.is_zero:
     return True
```

> I think that a lot of cases like this will have the same issue.

It is likely that some tests will need to be changed but first try to make the tests pass.

oscarbenjamin commented on sympy/core/add.py:
```diff
@@ -610,6 +610,21 @@ def as_numer_denom(self) -> tuple[Expr, Expr]:
 
         return _keep_coeff(ncon, n), _keep_coeff(dcon, d)
 
+    @property
+    def is_hermitian(self):
+        from sympy.simplify import simplify
+        if simplify(self.func(*self.args) - self._eval_adjoint()).is_zero:
+            return True
+        return
+
+    @property
+    def is_antihermitian(self):
+        from sympy.simplify import simplify
+        if simplify(self.func(*self.args) + self._eval_adjoint()).is_zero:
+            return True
+        return
```
Let's not have this change mixed in with the bigger change of removing hermitian from the assumptions system.

mathiscr commented:
Should I remove every predicates and queries that is linked to hermitian and anti hermitian in the PR that removes hermitian and anti hermitian in the assumptions system ?

oscarbenjamin commented:
> Should I remove every predicates and queries that are linked to hermitian and anti hermitian in the PR that removes hermitian and anti hermitian in the assumptions system ?

I don't understand this question.

mathiscr commented:
Sorry, I was asking if I should only 

> > Should I remove every predicates and queries that are linked to hermitian and anti hermitian in the PR that removes hermitian and anti hermitian in the assumptions system ?
> 
> I don't understand this question.

Sorry, I was a bit confused about what belonged to the old versus the new assumptions system. 

I created a new PR that removes hermitian and antihermitian from the old assumptions system.

TiloRC commented:
The new assumptions pretty much entirely contained in `sympy/assumptions`. So anything assumption related not in that folder is almost certainly related to the old assumptions.

mathiscr commented:
If #27997 gets merged, should I open a new PR based on it to implement what this PR (#27989) was originally intended to do ?

TiloRC commented:
Continuing this one or opening a new one are both fine as far as I'm aware. Depending on how comfortable you are with git, it might be easier to open a new one.

mathiscr commented:
I created a new PR: #28002 that improves only the `Mul` class. If it looks good for the `Mul` class, I plan to open two more PRs to improve `is_hermitian` and `is_antihermitian` for the `Add` and `Pow` classes.
